Scratch2

TimothyPMann opened the pull request:
Some changes prompted by my most recent exchanges with Rüdiger Kurth.

TimothyPMann committed:
Delete obsolete RCS "$Id ... $" lines

TimothyPMann committed:
Make dmk2cw warn if not all sector data fits on the physical track

The warning feature does not work with Catweasel MK1 or MK3, only MK4,
as it requires reading back the Catweasel memory pointer.

Add some hints to the dmk2cw man page on what to do if you get this
warning.

Bonus: To help support the feature (and because it generally seems
useful), cw2dmk now records the quirk flags in a previously
unused/reserved byte of the DMK file header.  dmk2cw now uses this
information to know that sectors are longer with certain quirks and
warn if the extra bytes didn't fit.  dmk2jv3 now warns if any quirk
bits are set, since JV3 format doesn't support any quirks. (Although
some quirks wouldn't really affect JV3, extra bytes from the various
QUIRK_EXTRA* quirks are lost in the conversion to JV3.)

TimothyPMann committed:
Check that numeric option arguments are valid numbers

Before this change, a sequence like "-i -T20,250 -h0" would be
silently treated as "-i 0 -h 0".  Now it results in the error message
"cw2dmk: -i requires a numeric argument".

qbarnes commented on cw2dmk.c:
```diff
@@ -1759,6 +1759,40 @@ parse_tracks(const char *nptr, int *rav)
+ * invalid characters or the string is empty.
+ */
+long int
+strtol_strict(const char *nptr, int base, const char name)
```
The `const` on `const char name` doesn't do anything for the function's API since `name` isn't a pointer.  It'd be the same as doing `const int base`.  But it doesn't hurt anything either.  Up to you if you leave it as is.

However, I'd recommend changing it to a `const char *name` (and adjusting the `fatal_msg` call).  We might eventually want to support long form command line arguments (e.g. `--verbosity`), so using a string arg could help support that eventuality.

TimothyPMann replied:
Changing it to take a string would make it harder to use, as the call sites only have a char. Hmm, well, I could create a string just above the switch (ch). That's not so bad.

(The unneeded const is left from my first draft code where the name argument actually was a const char*. I changed that almost immediately when I noticed the callers only had a char, but missed removing const.)

TimothyPMann replied:
I've taken this suggestion in 2c1018b

qbarnes commented on dmk2cw.c:
```diff
@@ -372,6 +372,24 @@ approx(int a, int b)
+
+  res = strtol(nptr, &endptr, base);
+  if (*nptr == '\0' || *endptr != '\0') {
+    fprintf(stderr, "dmk2cw: -%c requires a numeric argument\n", name);
```
This reminds me that `dmk2cw` still needs the `fatal_msg` and related functions too.  :/

TimothyPMann replied:
Yeah. I ran out of time or I might have cleaned that up too.

TimothyPMann replied:
I looked at that some today. There is one snag in factoring out those functions and calling them from dmk2cw as well: msg has some special treatment for OUT_RAW and OUT_HEX because unlike all the other levels, OUT_RAW is not cumulative -- it omits the output that OUT_HEX adds. This special treatment is only appropriate for cw2dmk, because dmk2cw and the other functions have different sets of verbosity levels.

It would work anyway because none of the others go up as far as 7 -- but, ugh.

So I'm not going to do this until I think of a nice way to abstract it out.

TimothyPMann committed:
Make the strto*_strict functions take a char*, not just a char

qbarnes commented on dmk2cw.c:
```diff
@@ -908,9 +937,9 @@ main(int argc, char** argv)
 	  if (out_fmt >= OUT_BYTES) {
 	    printf("%s%c%02x", out_fmt >= OUT_SAMPLES ? "\n" : "",
-		   "-FIAMJBX"[encoding], byte);
+		   encoding_letter[encoding], byte);
```
Don't you have to do here `encoding_letter[encoding & ~SECTOR_END]` or am I not following something?

TimothyPMann replied:
Good catch. I forgot to mask off the extra flag bit in several places.
I also noticed an opportunity to clean up some code that checks if the encoding is a subtype of FM versus MFM to decide how to fill extra gap areas that aren't in the DMK file, so I changed that too. I am testing this a bit more now...

TimothyPMann replied:
p.s. I've done some testing of writing and reading back a couple of different kinds of disks, including the -g option and -f/-l options to force extra gap fill at the start or end of tracks. It all seems to work. Will push this commit in a second.

qbarnes commented on cw2dmk.c:
```diff
@@ -1802,6 +1802,7 @@ main(int argc, char** argv)
   int cw_mk = 1;
   char *replay = NULL;
   FILE *replay_file = NULL;
+  char optname[3] = "-?";
```
Another way to have done this is just before calling `getopt()` save `optind` in a var (e.g. `ch_optind`) then instead of calling `strto.*strict()` with `optname`, use `argv[ch_optind]` (and at the bottom of the `while` loop, `ch_optind = optind;` to update).

TimothyPMann replied:
Does that work? getopt handles option arguments that are either part of the same argv[] entry or in the next one, such as -v1 versus -v 1. So if a user wrote -v1z, wouldn't the optind method complain that "-v1z requires a numeric argument" instead of "-v requires a numeric argument"? Even if it works for some reason, it seems more confusing.

qbarnes replied:
Oh, yes.  Silly oversight on my part not considering combined options without option-arguments.

TimothyPMann committed:
Fix a bug in efd81c6 where SECTOR_END wasn't always stripped out of encoding

Also cleans up code that generates extra gap bytes, using a common
macro to correctly test whether the prevailing encoding implies FM or
MFM gap bytes are needed.

qbarnes reviewed (commented):
I noticed the RCS strings are still present in *.man files.  Was this intentional or an oversight?

TimothyPMann commented:
Oversight. I forgot those were ever there.

On Tue, Aug 16, 2022 at 2:38 PM Quentin Barnes ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> I noticed the RCS strings are still present in *.man files. Was this
> intentional or an oversight?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/qbarnes/cw2dmk/pull/26#pullrequestreview-1074806879>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AILDEHKUI5VH7TWU3TDLVGLVZQC6VANCNFSM54QVHEHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

TimothyPMann committed:
Remove RCS Id strings from man pages too

qbarnes commented on catweasl.c:
```diff
@@ -187,6 +188,7 @@ unsigned char Mk3ControlBit[] = {
 #define MEMSIZE 131072
 #define CREG(c) (c->private[0])
 #define PTR(c) (c->private[1])
+#define SECTEND(c) (c->private[2])
```
I found `SECEND()` confusing not realizing at first that it implied a transitory state.  Would there be a better name for it to imply that nature, maybe `LASTSECEND()`?

qbarnes commented on cwfloppy.h:
```diff
@@ -60,6 +60,11 @@ int catweasel_read(catweasel_drive *d, int side, int clock, int time, int idx);
 
 /* Write data -- msdelay will be used */
 /* If time = 0, write from index hole to index hole */
+/* Return values:
```
Wish the function's descriptive comments were with the definition, not the declaration.  Once a declaration is placed in a header file, I would expect few devs would look in a header file for a function's comment  block about what the function does and/or returns.

qbarnes commented on dmk2cw.c:
```diff
@@ -358,15 +358,19 @@ approx(int a, int b)
+#define encoding_letter "-FIAMJBX"
+
+/* Or'ed into encoding at end of sector data */
+#define SECTOR_END 8
```
Not sure why SECTOR_END is in base 10 and made to appear in sequence with the other defines SKIP thorugh RX02.  SECTOR_END is a bitfield, so I'd prefer to see it in base 16.  Also, why not make SECTOR_END an MSB with a value of 0x80, that way if other encodings are added SECTOR_END doesn't have to be changed?

qbarnes commented on dmk2cw.c:
```diff
@@ -383,9 +387,11 @@ main(int argc, char** argv)
+  int sector_data;
   int cw_mk = 1;
   int tracklen;
+  int extra;
```
The variable `extra` could be named more descriptively.  Extra what?  Maybe `extra_bytes` or `quirk_bytes`?

qbarnes commented on dmk2cw.c:
```diff
@@ -813,13 +830,16 @@ main(int argc, char** argv)
-	  }
+	  } else {
+            /* Compute expected sector size */
+            sector_data = 2 + (128 << dmk_track[idamp+4]) + extra;
```
Why doesn't this line's computation match up with the implementation of `secsize()` in `cw2dmk.c`?

Is this the new sector end detection not going to work for non-IBM  FM formatted disks?  What about high order bits with MFM?

[Not related to this PR's changes] Does current `secsize()` implementation undercompute size when QUIRK_EXTRA_CRC and/or QUIRK_EXTRA are set?

qbarnes commented on dmk2cw.c:
```diff
@@ -813,13 +830,16 @@ main(int argc, char** argv)
-		    (encoding == FM && (dmk_track[datap-2] == 0x00 ||
-					dmk_track[datap-2] == 0xff)))) {
+		    (encoding == FM && (dmk_track[datap-fmtimes] == 0x00 ||
+					dmk_track[datap-fmtimes] == 0xff)))) {
```
Interesting the s/2/fmtimes/ bug hadn't turned up before.

qbarnes reviewed (commented):
Let me know what you think of this round.  I feel I'm completely done with reviewing these changes now.

I understand much better the changes you've made now.  Sorry for taking so long.

TimothyPMann commented:
Replies below.

On Mon, Aug 29, 2022 at 3:31 PM Quentin Barnes ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> Let me know what you think of this round. I feel I'm completely done with
> reviewing these changes now.
>
> I understand much better the changes you've made now. Sorry for taking so
> long.
> ------------------------------
>
> In catweasl.c
> <https://github.com/qbarnes/cw2dmk/pull/26#discussion_r957837579>:
>
> > @@ -187,6 +188,7 @@ unsigned char Mk3ControlBit[] = {
>  #define MEMSIZE 131072
>  #define CREG(c) (c->private[0])
>  #define PTR(c) (c->private[1])
> +#define SECTEND(c) (c->private[2])
>
> I found SECEND() confusing not realizing at first that it implied a
> transitory state. Would there be a better name for it to imply that nature,
> maybe LASTSECEND()?
>
I guess so. I don't mind changing it.

> ------------------------------
>
> In cwfloppy.h
> <https://github.com/qbarnes/cw2dmk/pull/26#discussion_r957838769>:
>
> > @@ -60,6 +60,11 @@ int catweasel_read(catweasel_drive *d, int side, int clock, int time, int idx);
>
>  /* Write data -- msdelay will be used */
>  /* If time = 0, write from index hole to index hole */
> +/* Return values:
>
> Wish the function's descriptive comments were with the definition, not the
> declaration. Once a declaration is placed in a header file, I would expect
> few devs would look in a header file for a function's comment block about
> what the function does and/or returns.
>
Yes, this is a somewhat odd style for C. IIRC, I inherited it from this
code that came from Michael Krause's Linux driver.
I suppose we could move all the comments in a separate changeset at some
point if one of us is motivated.

> ------------------------------
>
> In dmk2cw.c
> <https://github.com/qbarnes/cw2dmk/pull/26#discussion_r957839408>:
>
> > -#define MFM_IAM 5  /* MFM C2 with missing clock */
> -#define MFM_AM  6  /* MFM A1 with missing clock */
> -#define RX02    7  /* DEC-modified MFM as in RX02 */
> +/* Byte encodings and letters used when logging them */
> +#define SKIP    0  /* -  padding byte in FM area of a DMK */
> +#define FM      1  /* F  FM with FF clock */
> +#define FM_IAM  2  /* I  FM with D7 clock (IAM) */
> +#define FM_AM   3  /* A  FM with C7 clock (IDAM or DAM) */
> +#define MFM     4  /* M  MFM with normal clocking algorithm */
> +#define MFM_IAM 5  /* J  MFM C2 with missing clock */
> +#define MFM_AM  6  /* B  MFM A1 with missing clock */
> +#define RX02    7  /* X  DEC-modified MFM as in RX02 */
> +#define encoding_letter "-FIAMJBX"
> +
> +/* Or'ed into encoding at end of sector data */
> +#define SECTOR_END 8
>
> Not sure why SECTOR_END is in base 10 and made to appear in sequence with
> the other defines SKIP thorugh RX02. SECTOR_END is a bitfield, so I'd
> prefer to see it in base 16. Also, why not make SECTOR_END an MSB with a
> value of 0x80, that way if other encodings are added SECTOR_END doesn't
> have to be changed?
>
0x80 sounds reasonable. Changed.

> ------------------------------
>
> In dmk2cw.c
> <https://github.com/qbarnes/cw2dmk/pull/26#discussion_r957839899>:
>
> >    int cw_mk = 1;
>    int tracklen;
> +  int extra;
>
> The variable extra could be named more descriptively. Extra what? Maybe
> extra_bytes or quirk_bytes?
>
Either sounds good. I went with extra_bytes.

> ------------------------------
>
> In dmk2cw.c
> <https://github.com/qbarnes/cw2dmk/pull/26#discussion_r957842305>:
>
> > @@ -813,13 +830,16 @@ main(int argc, char** argv)
>  	    rx02_data = 2 + (256 << dmk_track[idamp+4]);
>  	    /* Follow CRC with one RX02-MFM FF */
>  	    dmk_track[fmtimes + datap + rx02_data++] = 0xff;
> -	  }
> +	  } else {
> +            /* Compute expected sector size */
> +            sector_data = 2 + (128 << dmk_track[idamp+4]) + extra;
>
> Why doesn't this line's computation match up with the implementation of
> secsize() in cw2dmk.c?
>
This computation has a "2 +" that counts the two CRC bytes, and...

> Is this the new sector end detection not going to work for non-IBM FM
> formatted disks? What about high order bits with MFM?
>
...yes, I see I didn't handle the rarely used "non-IBM" feature here, and I
didn't handle ignoring the high order bits on bogus sizecode fields that
are larger than maxsize (even more rarely used). Really, I should factor
out secsize from cw2dmk.c and use the same function as far as possible.
I'll look at changing that when I have time this weekend.

> [Not related to this PR's changes] Does current secsize() implementation
> undercompute size when QUIRK_EXTRA_CRC and/or QUIRK_EXTRA?
>
No, because secsize() in cw2dmk only needs to compute the number of bytes
covered by the first CRC. QUIRK_EXTRA_DATA adds 4 bytes to that, but the
other two quirks either cover the extra bytes with a separate CRC or don't
cover them with a CRC at all.
So the requirement in dmk2cw is a little different, but maybe more code can
be shared anyway, and certainly I can handle the weird rare cases just in
case.

> ------------------------------
>
> In dmk2cw.c
> <https://github.com/qbarnes/cw2dmk/pull/26#discussion_r957843578>:
>
> >  		   && !got_iam && dmk_track[datap] == 0xfc &&
>  		   ((encoding == MFM && dmk_track[datap-1] == 0xc2) ||
> -		    (encoding == FM && (dmk_track[datap-2] == 0x00 ||
> -					dmk_track[datap-2] == 0xff)))) {
> +		    (encoding == FM && (dmk_track[datap-fmtimes] == 0x00 ||
> +					dmk_track[datap-fmtimes] == 0xff)))) {
>
> Interesting the s/2/fmtimes/ bug hadn't turned up before.
>
Yes, i just happened to catch that one. Even the new code is a bit sleazy,
as datap could be 0, in which case we'd be looking at the two bytes prior
to the start of the track -- those are the last IDAM pointer, which is
effectively always 0. I felt bad about not cleaning that up further but
kind of ran out of time and energy.

> —
> Reply to this email directly, view it on GitHub
> <https://github.com/qbarnes/cw2dmk/pull/26#pullrequestreview-1089412987>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AILDEHOHSTODDYK6AYHTLL3V3U225ANCNFSM54QVHEHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

TimothyPMann committed:
Cleanups after review comments

secsize() is now moved to secsize.c and shared by cw2dmk and dmk2cw,
but each one uses separate definitions for encodings (FM, MFM, etc.),
so it is #included instead of separately compiled.  In the future,
perhaps this can be cleaned up further and more code can be shared.

TimothyPMann committed:
Make the -v2 and -v3 output of dmk2cw more readable.

See the man page for details.

TimothyPMann committed:
Update ChangeLog

qbarnes commented:
Everything looks good.  Yeah, we can let the MFM #define cleanup for another day.

Are all these commits ready to go and for me to crank out a new release with them?

TimothyPMann commented:
Yes, I think this stuff is ready for release. Thanks!

On Fri, Sep 16, 2022, 3:02 PM Quentin Barnes ***@***.***>
wrote:

> Everything looks good. Yeah, we can let the MFM #define cleanup for
> another day.
>
> Are all these commits ready to go and for me to crank out a new release
> with them?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/qbarnes/cw2dmk/pull/26#issuecomment-1249879598>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AILDEHKHKJBPIOJ6VYHY5H3V6TU53ANCNFSM54QVHEHA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
